lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb

```diff
@@ -20,9 +20,73 @@ def do_remove(name, type)
       send(method, name)
     end
 
+    # -1 = no conflict and create the record
+    #  0 = already exists and do nothing
+    #  1 = conflict and error out
+    def record_conflicts_ip(fqdn, type, address)
+      method = "ib_find_#{type.name.split('::').last.downcase}_record".to_sym
+      raise(Proxy::Dns::Error, "Finding of #{type} records not implemented") unless respond_to?(method, true)
+
+      return -1 if send(method, fqdn).empty?
+      return 0 if send(method, fqdn, address).any?
+      1
+    end
+
+    def record_conflicts_name(fqdn, type, content)
+      if type == Resolv::DNS::Resource::IN::PTR
+        record_conflicts_ip(content, type, fqdn)
+      else
+        record_conflicts_ip(fqdn, type, content)
+      end
+    end
+
     private
 
-    def ib_create_a_record(fqdn, address)
+    def ib_find_a_record(fqdn, address = nil)
+      params = {
+        :_max_results => 1,
+        :view => dns_view,
+        :name => fqdn
+      }
+      params[:ipv4addr] = address if address
+      Infoblox::Arecord.find(connection, params)
+    end
+
+    def ib_find_aaaa_record(fqdn, address = nil)
+      params = {
+        :_max_results => 1,
+        :view => dns_view,
+        :name => fqdn
+      }
+      params[:ipv6addr] = address if address
+      Infoblox::AAAArecord.find(connection, params)
+    end
+
+    def ib_find_ptr_record(fqdn, ptr = nil)
+      params = {
+        :_max_results => 1,
+        :view => dns_view,
+        :ptrdname => fqdn
+      }
+      if ptr
+        ip = IPAddr.new(ptr_to_ip(ptr))
+        params["ipv#{ip.ipv4? ? 4 : 6}addr".to_sym] = ip.to_s
+        params[:name] = ptr
+      end
+      Infoblox::Ptr.find(connection, params)
+    end
+
+    def ib_find_cname_record(fqdn, address = nil)
+      params = {
+        :_max_results => 1,
+        :view => dns_view,
+        :name => fqdn
+      }
+      params[:canonical] = address if address
+      Infoblox::Cname.find(connection, params)
+    end
+
+    def ib_create_a_record(fqdn, address = nil)
```

Comment:
When is the address nil? That doesn't sound like a valid use case

Reply:
Look at the `record_conflicts_ip` implementation, I could not figure out better way but this is how it works:

* It calls `ib_create_a_record("fqdn.example.com") first grabbing all records of the specific type. If that's empty it returns -1 (no conflict for sure).
* Then calls `ib_create_a_record("fqdn.example.com", "1.2.3.4") grabbing all records of the specific type with the specific address. If there is any record that means the record already exists and it returns 0.
* Otherwise it means there are some records with "fqdn.example.com" which are however not set to "1.2.3.4", therefore it returns 1 (conflict).

I know its a little bit lame, but I could not figure out how to do "return me fqdn.example.com A records which don't have value set to 1.2.3.4". This does it, it's just two simple queries both returning just empty array or array with maximum of one record. Alternatively I could do just one query grabbing ALL records of the particular type and fqdn and then iterating them, but possibly this could return many values (Infoblox entries also have lots of fields). I wanted to avoid that.

So back to the concern - all these helper methods can be called with or without 2nd parameter. It's also the reason why I swap params for the PTR record because it's actually swapped by the common code and only last parameter(s) are optional in Ruby.

Reply:
That calls `ib_find_a_record` and calling that without an address makes sense but this is the create method. Creation without an address sounds invalid.

I will also agree that the API on this is slightly inconsistent and confusing. I'd have preferred a `name, value` pair where the name is the full DNS name and the value whatever the DNS value for that record is. However, this is what we have now.

Reply:
Oh this was added as an accident! Optional arguments are only valid for `ib_find_*` methods, this was not intended.